src/sql/parts/notebook/models/notebookModel.ts

```diff
@@ -400,25 +411,84 @@ export class NotebookModel extends Disposable implements INotebookModel {
 				this._activeConnection = undefined;
 			}
 
-			clientSession.initialize();
+			await clientSession.initialize();
 			this._sessionLoadFinished = clientSession.ready.then(async () => {
 				if (clientSession.isInErrorState) {
 					this.setErrorState(clientSession.errorMessage);
 				} else {
 					this._onClientSessionReady.fire(clientSession);
 					// Once session is loaded, can use the session manager to retrieve useful info
-					this.loadKernelInfo(clientSession);
+					this.loadKernelInfo(clientSession, this.defaultKernel.display_name);
 				}
 			});
-		});
+		}
+	}
+
+	// When changing kernel, update the active session and register the kernel change event
+	// So KernelDropDown could get the event fired when added listerner on Model.KernelChange
+	private updateActiveClientSession(clientSession: IClientSession) {
+		this.clearClientSessionListeners();
+		this._activeClientSession = clientSession;
+		this._clientSessionListeners.push(this._activeClientSession.kernelChanged(e => this._kernelChangedEmitter.fire(e)));
+	}
+
+	private clearClientSessionListeners() {
+		this._clientSessionListeners.forEach(listener => listener.dispose());
+		this._clientSessionListeners = [];
+	}
+
+	public setDefaultKernelAndProviderId() {
+		if (this._savedKernelInfo) {
```

Comment:
Minor:
```
let kernelPriority = [this._savedKernelInfo, this._defaultKernel, sqlKernelSpec]
for (let i = 0; i < kernelPriority.length && !this._providerId; i++) {
    let kernel = kernelPriority[i];
    if (!kernel) { continue; }
    // otherwise get the providerId?
...
}